tests/subsys/settings/functional/src/CMakeLists.txt

```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_library_named(  settings_func_test)
+zephyr_library_sources(settings_basic_test.c)
+add_dependencies(      settings_func_test offsets_h)
```

Comment:
Why is this dependency required?

I see this in the top level zephyr/CMakeLists.txt:

```
foreach(zephyr_lib ${ZEPHYR_LIBS_PROPERTY})
  # TODO: Could this become an INTERFACE property of zephyr_interface?
  add_dependencies(${zephyr_lib} ${OFFSETS_H_TARGET})
endforeach()
```

So it looks like zephyr_library_named() should be taking care of this:

```
macro(zephyr_library_named name)
[...]
  set(ZEPHYR_CURRENT_LIBRARY ${name})
  add_library(${name} STATIC "")

  zephyr_append_cmake_library(${name})
[...]
endmacro()

[...]

function(zephyr_append_cmake_library library)
  set_property(GLOBAL APPEND PROPERTY ZEPHYR_LIBS ${library})
endfunction()
```

Reply:
I'm not sure why this explicit dependency is needed. I was suggested by @SebastianBoe . See below how `make -C build` always fails without it.

Note: `make -j` sometimes succeeds in building even without this dependency thanks to a race condition. `make -k; make` obviously succeeds too.

```
make: Entering directory 'zephyr/tests/subsys/settings/functional/nvs/build'
/usr/bin/cmake -Szephyr/tests/subsys/settings/functional/nvs -Bzephyr/tests/subsys/settings/functional/nvs/build --check-build-system CMakeFiles/Makefile.cmake 0
/usr/bin/cmake -E cmake_progress_start zephyr/tests/subsys/settings/functional/nvs/build/CMakeFiles zephyr/tests/subsys/settings/functional/nvs/build/CMakeFiles/progress.marks
make -f CMakeFiles/Makefile2 all
make[1]: Entering directory 'zephyr/tests/subsys/settings/functional/nvs/build'
make -f func_test_bindir/CMakeFiles/settings_func_test.dir/build.make func_test_bindir/CMakeFiles/settings_func_test.dir/depend
make[2]: Entering directory 'zephyr/tests/subsys/settings/functional/nvs/build'
cd zephyr/tests/subsys/settings/functional/nvs/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" zephyr/tests/subsys/settings/functional/nvs zephyr/tests/subsys/settings/functional/src zephyr/tests/subsys/settings/functional/nvs/build zephyr/tests/subsys/settings/functional/nvs/build/func_test_bindir zephyr/tests/subsys/settings/functional/nvs/build/func_test_bindir/CMakeFiles/settings_func_test.dir/DependInfo.cmake --color=
make[2]: Leaving directory 'zephyr/tests/subsys/settings/functional/nvs/build'
make -f func_test_bindir/CMakeFiles/settings_func_test.dir/build.make func_test_bindir/CMakeFiles/settings_func_test.dir/build
make[2]: Entering directory 'zephyr/tests/subsys/settings/functional/nvs/build'
[  0%] Building C object func_test_bindir/CMakeFiles/settings_func_test.dir/settings_basic_test.c.obj
In file included from zephyr/lib/libc/minimal/include/errno.h:20,
                 from zephyr/include/kernel.h:18,
                 from zephyr/include/zephyr.h:18,
                 from zephyr/tests/subsys/settings/functional/src/settings_basic_test.c:12:
zephyr/include/sys/errno_private.h:34:10: fatal error: syscalls/errno_private.h: No such file or directory
 #include <syscalls/errno_private.h>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [func_test_bindir/CMakeFiles/settings_func_test.dir/build.make:63: func_test_bindir/CMakeFiles/settings_func_test.dir/settings_basic_test.c.obj] Error 1
make[2]: Leaving directory 'zephyr/tests/subsys/settings/functional/nvs/build'
make[1]: *** [CMakeFiles/Makefile2:2972: func_test_bindir/CMakeFiles/settings_func_test.dir/all] Error 2
make[1]: Leaving directory 'zephyr/tests/subsys/settings/functional/nvs/build'
make: *** [Makefile:84: all] Error 2
make: Leaving directory 'zephyr/tests/subsys/settings/functional/nvs/build'
```

Reply:
That seems to be papering over a bug in the cmake extensions somehow, then. @SebastianBoe any comment?

I guess as a band-aid that makes sense then.

Reply:
`ninja -j1` doesn't fail, it seems "lucky" enough to build in some order that avoids the missing dependency issue.

However `ninja settings_func_test` fails with the same error than `make settings_func_test` above.

Reply:
@mbolivar It seems to be that at the time the foreach loop is executed the library settings_func_test has not been added to the list yet. I think the TODO statement in the foreach loop would fix that issue.

Reply:
@joerchan has it right. As do you @mbolivar , I'll look into moving the 'add_dependencies' call from the loop in zephyr/CMakeLists.txt to extensions.cmake.

Reply:
But for now this is best-practice.

Reply:
a black-magic it is

Reply:
Any sufficiently obscure code is indistinguishable from magic ...